File: kakao.app/js/ked/edit/tool/salt.js
var _k_ = {in: function (a,l) {return (typeof l === 'string' && typeof a === 'string' && a.length ? '' : []).indexOf.call(l,a) >= 0}, empty: function (l) {return l==='' || l===null || l===undefined || l!==l || typeof(l) === 'object' && Object.keys(l).length === 0}, lpad: function (l,s='',c=' ') {s=String(s); while(s.length<l){s=c+s} return s}}

var salt

import kxk from "../../../kxk.js"
let kseg = kxk.kseg

import salter from "../../../kxk/salter.js"


salt = (function ()
{
    function salt ()
    {}

    salt["isSaltedLine"] = function (line)
    {
        var trimmed

        trimmed = kseg.trim(kseg.trim(kseg.trim(line),'#'))
        if (_k_.in(trimmed[0],'█0'))
        {
            return kseg.collectGraphemes(trimmed).length <= 3
        }
    }

    salt["findPositionsForSaltInsert"] = function (lines, pos)
    {
        var ey, posl, sy, y

        y = pos[1]
        if (!this.isSaltedLine(lines[y]))
        {
            return
        }
        sy = y
        while (this.isSaltedLine(lines[sy - 1]))
        {
            sy -= 1
            if (y - sy >= 4)
            {
                break
            }
        }
        ey = y
        while (this.isSaltedLine(lines[ey + 1]))
        {
            ey += 1
            if (ey - sy >= 4)
            {
                break
            }
        }
        posl = []
        if (ey - sy >= 4)
        {
            for (var _a_ = y = sy, _b_ = sy + 4; (_a_ <= _b_ ? y <= sy + 4 : y >= sy + 4); (_a_ <= _b_ ? ++y : --y))
            {
                posl.push([pos[0],y])
            }
        }
        return posl
    }

    salt["insertAsciiHeaderForPositionsAndRanges"] = function (lines, posl, ranges)
    {
        var indt, salt, text

        if (_k_.empty(ranges))
        {
            ranges = posl.map((function (p)
            {
                return this.rangeOfClosestWordToPos(lines,p)
            }).bind(this))
        }
        text = this.joinLines(this.textForLineRanges(lines,ranges),' ')
        indt = _k_.lpad(this.lineIndentAtPos(lines,posl[0]))
        salt = salter(text,{prepend:indt + '# '}) + '\n'
        var _a_ = this.insertTextAtPositions(lines,salt,[[0,posl[0][1]]]); lines = _a_[0]; posl = _a_[1]

        return [lines,posl,[]]
    }

    return salt
})()

export default salt;

File: kakao.app/love/lua/edit/tool/salt.lua
--[[
 0000000   0000000   000      000000000
000       000   000  000         000   
0000000   000000000  000         000   
     000  000   000  000         000   
0000000   000   000  0000000     000   
--]]

-- use ../../../kxk ▪ kseg
-- use ../../../kxk ◆ salter


local salt = class("salt")
    

--  0000000   0000000   000      000000000  00000000  0000000    
-- 000       000   000  000         000     000       000   000  
-- 0000000   000000000  000         000     0000000   000   000  
--      000  000   000  000         000     000       000   000  
-- 0000000   000   000  0000000     000     00000000  0000000    


function salt.static.isSaltedLine(line) 
        local trimmed = kseg.trim(kseg.trim(kseg.trim(line), '#'))
        if (trimmed[0] in '█0') then 
            return (#kseg.collectGraphemes(trimmed) <= 3)
        end
    end

--  0000000   0000000   000      000000000        000  000   000   0000000  00000000  00000000   000000000  
-- 000       000   000  000         000           000  0000  000  000       000       000   000     000     
-- 0000000   000000000  000         000           000  000 0 000  0000000   0000000   0000000       000     
--      000  000   000  000         000           000  000  0000       000  000       000   000     000     
-- 0000000   000   000  0000000     000           000  000   000  0000000   00000000  000   000     000     


function salt.static.findPositionsForSaltInsert(lines, pos) 
        local y = pos[1]
        if not salt.isSaltedLine(lines[y]) then return end
        
        local sy = y
        while salt.isSaltedLine(lines[(sy - 1)]) do 
            sy = sy - 1
            if ((y - sy) >= 4) then break end
        end
        
        local ey = y
        while salt.isSaltedLine(lines[(ey + 1)]) do 
            ey = ey + 1
            if ((ey - sy) >= 4) then break end
        end
        
        local posl = array()
        if ((ey - sy) >= 4) then 
            for y in iter(sy, (sy + 4)) do 
                posl.push(array(pos[0], y))
            end
        end
        
        return posl
    end

--  0000000    0000000   0000000  000  000  000   000  00000000   0000000   0000000    00000000  00000000   
-- 000   000  000       000       000  000  000   000  000       000   000  000   000  000       000   000  
-- 000000000  0000000   000       000  000  000000000  0000000   000000000  000   000  0000000   0000000    
-- 000   000       000  000       000  000  000   000  000       000   000  000   000  000       000   000  
-- 000   000  0000000    0000000  000  000  000   000  00000000  000   000  0000000    00000000  000   000  


function salt.static.insertAsciiHeaderForPositionsAndRanges(lines, posl, ranges) 
        if empty(ranges) then 
            ranges = posl.map(method (p) 
    return salt.rangeOfClosestWordToPos(lines, p)
end)
        end
        
        local text = salt.joinLines(salt.textForLineRanges(lines, ranges), ' ')
        
        local indt = lpad(salt.lineIndentAtPos(lines, posl[0]))
        
        local salt = (salter(text, (prepend:indt + '# ')) + '\n')
        
        array(lines, posl) = salt.insertTextAtPositions(lines, salt, array(array(0, posl[0][1])))
        return array(lines, posl, array())
    end

return salt